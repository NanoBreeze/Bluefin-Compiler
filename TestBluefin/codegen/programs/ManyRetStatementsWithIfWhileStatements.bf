/*
Verifies multiple return statements, and that code after a return statement will be 
put in its own BasicBlock (to ensure each BasicBlocks will have more only one terminator
and it's at the end)
*/

void foo(int a) {

	a + 6;
	return; // every line after this is unreachable, technically. Create a new BasicBlock right after this return

	if (a == 1) {
		return;
	}
	else {
		return;
	}
}

int wow(int a) {
	while (a > 1) {
		if (a > 2) {
			return 2;
		}
		else {
			return 3;
		}
		return 6; // yes, we accept that this "merge" BB will have to 'callers'/predecessors
		// because 'return 2' and 'return 3' will create an "impossible" BB, into which the branching
		// instructions to "merge" are placed. But since "impossible" BBs are removed from code generation
		// nothing can call this "merge" BB
	}

	return 7;
	return 8;
	a + 9;
	return 10;
}