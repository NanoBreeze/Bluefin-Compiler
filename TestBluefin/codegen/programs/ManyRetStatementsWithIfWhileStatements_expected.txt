; ModuleID = 'ManyRetStatementsWithIfWhileStatements.bf'
source_filename = "ManyRetStatementsWithIfWhileStatements.bf"

define void @foo(i32 %a) {
entry:
  %alloctmp_a = alloca i32, align 4
  store i32 %a, i32* %alloctmp_a, align 4
  %loadtmp_a = load i32, i32* %alloctmp_a, align 4
  %addtmp = add i32 %loadtmp_a, 6
  ret void

then:                                             ; No predecessors!
  ret void

else:                                             ; No predecessors!
  ret void

merge:                                            ; No predecessors!
  ret void
}

define i32 @wow(i32 %a) {
entry:
  %alloctmp_a = alloca i32, align 4
  store i32 %a, i32* %alloctmp_a, align 4
  br label %whileLoopCmp

whileLoopCmp:                                     ; preds = %entry
  %loadtmp_a = load i32, i32* %alloctmp_a, align 4
  %cmpSGTtmp = icmp sgt i32 %loadtmp_a, 1
  br i1 %cmpSGTtmp, label %whileLoopBody, label %afterWhileLoop

whileLoopBody:                                    ; preds = %whileLoopCmp
  %loadtmp_a1 = load i32, i32* %alloctmp_a, align 4
  %cmpSGTtmp2 = icmp sgt i32 %loadtmp_a1, 2
  br i1 %cmpSGTtmp2, label %then, label %else

afterWhileLoop:                                   ; preds = %whileLoopCmp
  ret i32 7

then:                                             ; preds = %whileLoopBody
  ret i32 2

else:                                             ; preds = %whileLoopBody
  ret i32 3

merge:                                            ; No predecessors!
  ret i32 6
}
